raft-tech/TANF-app
# 2795 s1 duplicates

## Summary of Changes
- Added new class that manages duplicate detection while case cat4 is occurring
- Updated cat4 validation to leverage duplicate detector
- Updated parse logic to not create records in a case if it has any cat4 errors
Pull request closes #2795 

## CPU and Memory Analysis
- Used Docker to track the `web` container stats during parsing of the largest file we have in the repo (~50MB) on the `develop` branch and this branch. 
- The overall results can be found in `super_big_file_results.txt`. The data listed in that text file can be found in the accompanying files: `develop_super_big_file.txt` and `dup_super_big_file.txt`.
- The data was generated with: `track_docker_stats.sh.txt` (GitHub won't let you upload a pure shell file).
[super_big_file_results.txt](https://github.com/user-attachments/files/15534179/super_big_file_results.txt)
[dup_super_big_file.txt](https://github.com/user-attachments/files/15534181/dup_super_big_file.txt)
[develop_super_big_file.txt](https://github.com/user-attachments/files/15534183/develop_super_big_file.txt)
[track_docker_stats.sh.txt](https://github.com/user-attachments/files/15534202/track_docker_stats.sh.txt)


## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Create a file for TANF section 1 - 4 that has one or more duplicate or partial duplicate records.
3. Submit file
4. Verify no records tied to that case are created and the error report reflects cat4 duplicate errors

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] In section 1 files, exact and partial duplicate records are detected, not stored in db, and error message included in the feedback report.
- [x] In section 2 files, exact and partial duplicate records are detected, not stored in db, and error message included in the feedback report.
- [x] section 3 files, partial duplicate T6 records are detected, not stored in db, and error message included in the feedback report.
- [x] section 4 files, partial duplicate T7 records are detected, not stored in db, and error message included in the feedback report.
- [x] duplicate detection logic is applied consistently to TANF, SSP, and Tribal TANF files. 
- [x] cat 4 records not stored in db

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)
